Classes/common/CDTIncrementalStore.m

```diff
@@ -0,0 +1,2409 @@
+        [[CDTIndexManager alloc] initWithDatastore:datastore error:&err];
+    if (!indexManager) {
+        if (error) *error = err;
+        CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"%@: %@: Cannot create indexManager: %@", CDTISType,
+                    databaseName, err);
+        return NO;
+    }
+
+    // Commit before setting up replication
+    self.databaseName = databaseName;
+    self.datastore = datastore;
+    self.manager = manager;
+    self.indexManager = indexManager;
+
+    if (![self setupReplicators:remoteURL manager:manager datastore:datastore]) {
+        CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"%@: continuing without replication", CDTISType);
+    }
+
+    return YES;
+}
+
+/**
+ *  Setup the indexes.
+ *  We only care about one right now and that is to find an entity name
+ *
+ *  @param error Error
+ *
+ *  @return YES/NO
+ */
+- (BOOL)setupIndexes:(NSError **)error
+{
+    NSError *err = nil;
+
+    // the big index
+    if (![self.indexManager ensureIndexedWithIndexName:CDTISEntityNameKey
+                                             fieldName:CDTISEntityNameKey
+                                                 error:&err]) {
+        if (error) *error = err;
+        CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"%@: %@: cannot create default index: %@", CDTISType,
+                    self.databaseName, err);
+        return NO;
+    }
+
+    return YES;
+}
+
+/**
+ *  Encode version hashes, which come to us as a dictionary of inline data
+ *  objects, so we encode them as a hex string.
+ *
+ *  @param hashes hashes
+ *
+ *  @return encoded dictionary
+ */
+static NSDictionary *encodeVersionHashes(NSDictionary *hashes)
+{
+    NSMutableDictionary *newHashes = [NSMutableDictionary dictionary];
+    for (NSString *hash in hashes) {
+        NSData *h = hashes[hash];
+        NSString *s = CDTISStringFromData(h);
+        newHashes[hash] = s;
+    }
+    return [NSDictionary dictionaryWithDictionary:newHashes];
+}
+
+/**
+ *  The version hashes are inline `NSData` and so they need to be encoded
+ *  We may have to consider that Apple may have other non encodable items
+ *  other than the hashes.
+ *
+ *  @param metadata metadata from Core Data
+ *
+ *  @return encoded dictionary
+ */
+static NSDictionary *encodeCoreDataMeta(NSDictionary *inMetaData)
+{
+    NSDictionary *hashes = inMetaData[NSStoreModelVersionHashesKey];
+    NSMutableDictionary *metaData = [NSMutableDictionary dictionary];
+
+    if (CDTISStoreAllCoreDataMetaData == NO) {
+        metaData[NSStoreUUIDKey] = inMetaData[NSStoreUUIDKey];
+        metaData[NSStoreTypeKey] = inMetaData[NSStoreTypeKey];
+    } else {
+        metaData = [inMetaData mutableCopy];
+    }
+
+    // hashes are inline data and need to be converted
+    if (hashes) {
+        metaData[NSStoreModelVersionHashesKey] = encodeVersionHashes(hashes);
+    }
+    return [NSDictionary dictionaryWithDictionary:metaData];
+}
+
+- (NSDictionary *)updateObjectModel
```

Comment:
Does implicit migration de-dup the older objects? @pbohrer recalls no duplicates.